chrome-extension/course_sort/course_sort.js

```diff
@@ -0,0 +1,242 @@
+      courseListResults.firstElementChild.appendChild(courseDiv);
+    });
+
+    document.querySelector('.size-dropdown').value = dropdownValue;
+  }).catch(error => {
+    console.error(courseSortingLogPrefix + 'Error building course dictionary:', error);
+  });
+};
+
+const sortCoursesTemplate = (buildDictFunc, additionalLogic) => {
+  return buildDictFunc().then(courseDict => {
+    additionalLogic();
+    return courseDict;
+  });
+};
+
+let courseSizeDict = {};
+
+const buildSeatsAvailableDict = () => {
+  return new Promise((resolve, reject) => {
+    let courseListResults = document.querySelector('.course-list-results');
+    let courseDivs = Array.from(courseListResults.firstElementChild.children);
+
+    if (Object.keys(courseSizeDict).length === 0) {
+      for (let courseDiv of courseDivs) {
+        let courseName = courseDiv.querySelector('a').getAttribute('name');
+        let targetDiv = courseDiv.querySelector('div:nth-child(2) > div:nth-child(3)');
+        let trElements = targetDiv.querySelectorAll('tr');
+  
+        let sum = 0;
+        trElements.forEach(tr => {
+          let lastChildText = tr.lastElementChild.innerText;
+  
+          // Merged extractDifference logic
+          let matches = lastChildText.match(/(\d+)\s+of\s+(\d+)/g);
+          if (matches) {
+            matches.forEach(match => {
+              let parts = match.match(/(\d+)\s+of\s+(\d+)/);
+              let current = parseInt(parts[1]);
+              let total = parseInt(parts[2]);
+              sum += (total - current);
+            });
+          }
+        });
+  
+        courseSizeDict[courseName] = sum;
+      }
+    }
+    
+    const hideSectionButtons = document.querySelectorAll('.action-sections.btn.btn-default.pull-right');
+    hideSectionButtons.forEach(button => button.click());
+    document.querySelector('.loadingIndicator').remove();
+
+    resolve(courseSizeDict);
+  });  
+};
+
+
+let courseMostCommonGradeDict = {};
+let courseAvgDict = {};
+let coursePopularityDict = {};
+let courseUnitsDict = {};
+
+const buildDictionary = (fetchDataFunc, dictToUpdate) => {
+  const loadingElement = `<h2 class='loadingIndicator'><p>Sorting...</p></h2>`;
+  let courseListResults = document.querySelector('.course-list-results').parentElement;
+  courseListResults.insertBefore(htmlToElement(loadingElement), courseListResults.firstChild);
+
+  return new Promise((resolve, reject) => {
+    let courseListResults = document.querySelector('.course-list-results');
+    let courseDivs = Array.from(courseListResults.firstElementChild.children);
+    let fetchPromises = [];
+
+    if (Object.keys(dictToUpdate).length === 0) {
+      for (let courseDiv of courseDivs) {
+        let courseName = courseDiv.querySelector('a').getAttribute('name');
+        let fetchPromise = new Promise((resolve, reject) => {
+          try {
+            // For debuggging, use `https://gophergrades-9wharkgah-umn-social-coding.vercel.app/api/class/${courseName}`
```

Comment:
this comment can be removed because as soon as this PR is merged we won't need to do this again